pkg/ipam/crd_eni.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ipam
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	eniTypes "github.com/cilium/cilium/pkg/aws/eni/types"
+
+	ciliumv2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/sirupsen/logrus"
+
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+)
+
+type linkMap map[string]netlink.Link // by MAC addr
+
+func updateENIRulesAndRoutes(oldNode, newNode *ciliumv2.CiliumNode, mtuConfig MtuConfiguration) error {
+	eniByName := newNode.Status.ENI.DeepCopy().ENIs
+	expectedENIByMac := make(map[string]string)
+	firstInterfaceIndex := *newNode.Spec.ENI.FirstInterfaceIndex
+	for name, eni := range eniByName {
+		if eni.Number < firstInterfaceIndex {
+			continue
+		}
+		expectedENIByMac[eni.MAC] = name
+	}
+
+	newENIs, _ := diffResources(oldNode, newNode)
+	go setupENI(newENIs, expectedENIByMac, eniByName, firstInterfaceIndex, mtuConfig)
```

Comment:
There's no synchronization around this goroutine. In theory this might cause problems if `updateENIRulesAndRoutes` is called multiple times in quick succession, which could cause multiple `setupENI` goroutines to run concurrently, possibly stepping on each other's toes. In practice is should be OK though. 🤞

Reply:
This is a great point. I have reworks the PR to only consider newly added interfaces. This should ensure that all possibly concurrent go routine only try to configure a distinct subset of newly added ENIs.

In addition to this change, I also increased the wait time and introduced exponential backoff.